fearless_simd/src/lib.rs

```diff
@@ -18,21 +19,82 @@ pub use traits::*;
 // For now, only bring in f16 on aarch64. We can also bring it in
 // on x86_64, but only Sapphire Rapids supports it.
 
-#[cfg(target_arch = "aarch64")]
-pub mod aarch64;
-
-#[cfg(target_arch = "aarch64")]
-pub use aarch64::Level;
-
-#[cfg(target_arch = "x86_64")]
-pub mod x86_64;
-
-#[cfg(target_arch = "x86_64")]
-pub use x86_64::Level;
-
 #[cfg(all(target_arch = "aarch64", feature = "half"))]
 pub type f16 = half::f16;
 #[cfg(all(target_arch = "aarch64", not(feature = "half")))]
 mod half_assed;
 #[cfg(all(target_arch = "aarch64", not(feature = "half")))]
 pub use half_assed::f16;
+
+#[cfg(all(not(feature = "libm"), not(feature = "std")))]
+compile_error!("fearless_simd requires either the `std` or `libm` feature");
+
+#[cfg(all(feature = "std", target_arch = "aarch64"))]
+pub mod aarch64 {
+    pub use crate::generated::Neon;
+}
+
+/// The level enum with the specific SIMD capabilities available.
+#[derive(Clone, Copy, Debug)]
+pub enum Level {
+    Fallback(Fallback),
+    #[cfg(all(feature = "std", target_arch = "aarch64"))]
+    Neon(Neon),
+}
+
+impl Level {
+    #[cfg(all(feature = "std", target_arch = "aarch64"))]
+    pub fn new() -> Self {
+        if std::arch::is_aarch64_feature_detected!("neon") {
+            unsafe { Level::Neon(Neon::new_unchecked()) }
+        } else {
+            Self::fallback()
+        }
+    }
+
+    #[cfg(all(feature = "std", target_arch = "x86_64"))]
+    pub fn new() -> Self {
+        Self::fallback()
+    }
+    
+    #[cfg(any(not(feature = "std"), not(any(target_arch = "aarch64", target_arch = "x86_64"))))]
+    pub fn new() -> Self {
+        Self::fallback()
+    }
+
+    #[cfg(all(feature = "std", target_arch = "aarch64"))]
+    #[inline]
+    pub fn as_neon(self) -> Option<Neon> {
+        match self {
+            Level::Neon(neon) => Some(neon),
+            _ => None,
+        }
+    }
```

Comment:
I'm interested in your thoughts. My rough impression was that `fearless_simd` APIs such as `as_neon` act as a witness and thus are a helper for consumer libraries. E.g. in the `examples/play.rs`:

```rs
fn do_something_on_neon(_level: Level) -> f32 {
    #[cfg(target_arch = "aarch64")]
    if let Some(neon) = _level.as_neon() {
        return neon.vectorize(
            #[inline(always)]
            || {
                let v = neon.neon.vdupq_n_f32(42.0);
                neon.neon.vgetq_lane_f32::<0>(v)
            },
        );
    }
    0.0
}
```

Thus, my main question is, is it a better API for the method `as_neon` to completely vanish given `cfg`, or, have the method always present, but make it unconditionally return `None` if `not(all(feature = "std", target_arch = "aarch64"))`, maybe something like:

```suggestion
    #[inline]
    pub fn as_neon(self) -> Option<Neon> {
        match self {
            #[cfg(all(feature = "std", target_arch = "aarch64"))]
            Level::Neon(neon) => Some(neon),
            _ => None,
        }
    }
````